afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+            ptr::copy_nonoverlapping(iface.as_ref().as_ptr(),
+                ifr.ifr_ifrn.ifrn_name.as_mut_ptr(), libc::IFNAMSIZ);
+            // ioctl(SIOCGIFINDEX) fills in the index field of the ifreq object
+            // Resources:
+            // man 7 netdevice
+            let err = libc::ioctl(self.fd, linux::SIOCGIFINDEX, &ifr);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+
+            // bind the socket
+            let mut ll : libc::sockaddr_ll = MaybeUninit::zeroed().assume_init();
+            ll.sll_family = libc::AF_PACKET as libc::c_ushort;
+            ll.sll_ifindex = ifr.ifr_ifru.ifru_ivalue; // expanded from `ifr_ifindex` in kernel headers
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#bind
+            // man 7 packet regarding sockaddr_ll
+            let err = libc::bind(self.fd,
+                                 &mut ll as *mut _ as *mut libc::sockaddr,
+                                 mem::size_of::<libc::sockaddr_ll>() as libc::c_uint);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            ll
+        };
+        let fd = self.fd;
+        mem::forget(self);
+        Ok(BoundSocket { fd, send_addr })
+    }
+
+    /// Configures the socket's non-blocking status.
+    pub fn set_nonblocking(&mut self, nonblocking: bool) -> io::Result<()> {
+        // This block is unsafe because it requires syscalls, but it is safe because each syscall's
+        // error status is checked, and it does not interact with Rust-held memory.
+        // Resources used to write syscall code:
+        // https://beej.us/guide/bgnet/html/multi/advanced.html#blocking
+        // man 2 fcntl
+        unsafe {
+            let flags = libc::fcntl(self.fd, libc::F_GETFL);
+            if flags < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            let new_flags = if nonblocking {
+                flags | libc::O_NONBLOCK
+            } else {
+                flags & (!libc::O_NONBLOCK)
+            };
+            let err = libc::fcntl(self.fd, libc::F_SETFL, new_flags);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+        }
+        Ok(())
+    }
+
+    /// Returns true if the socket is configured not to block, false otherwise.
+    pub fn is_nonblocking(&self) -> io::Result<bool> {
+        // See comments on block above (in set_nonblocking).
+        let flags = unsafe {
+            let flags = libc::fcntl(self.fd, libc::F_GETFL);
+            if flags < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            flags
+        };
+        Ok(flags & libc::O_NONBLOCK == libc::O_NONBLOCK)
+    }
+}
+
+impl BoundSocket {
+    /// Sends a frame to the NIC.
+    pub fn send(&mut self, frame: &[u8]) -> io::Result<usize> {
```

Comment:
Is the idea here that clients of `BoundSocket` will just serialize whatever data type they're working with to a slice of bytes to be sent? I like the interface since it's the most flexible, but it means we should include serialization helpers in the `smoltcp` types we're using in `runtime` if that's necessary.

Reply:
Yes. Just serialize the data into a slice. I'd rather not depend on `smoltcp` here.